### PR TITLE
tmpop: add tendermint evidence

### PR DESCRIPTION
**Pull request by t-bast**
At last, we're able to compute full tendermint evidence for our segments!
We can't really use the lite package as it's meant to be used by clients and is quite complex to fit in our infrastructure (it acts as a proxy node) but I think the data we're able to access in the votes is enough to produce a correct proof.
It's a bit complicated to do so because we can do it only at the beginning of block N+3 so we need to be careful with off-by-one errors in many places.
I've tested on a filetmpop that the evidence produced looks correct and is correctly validated by the Verify() method, but this is a tricky piece of code so I'm counting on you to find potential issues during the PR ;).
Don't hesitate to drop by my desk to draw diagrams of what object signs what part if you're unsure.

---
This change is [Reviewable](https://reviewable.io/reviews/stratumn/go-indigocore/356)

### CODE REVIEW

**Comment by such** [comment]
:lgtm:

---

Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[cs/evidence_tendermint_test.go, line 189 at r1](https://beta.reviewable.io/reviews/stratumn/go-indigocore/356#-L6vuIIG3kW5pP9uI1Oa:-L6vuIIG3kW5pP9uI1Ob:b-6fi506) ([raw file](https://github.com/stratumn/go-indigocore/blob/390d13ae61c452d70ba949191b3cfe7f15e0555e/cs/evidence_tendermint_test.go#L189)):*
> ```Go
> 	}
> 
> 	assert.True(t, e.Verify(linkHash), "Proof should be verified")
> ```

supernit: shouldn't this assert be in its own test case?

---

*[tmpop/tmpop.go, line 376 at r1](https://beta.reviewable.io/reviews/stratumn/go-indigocore/356#-L6vwmRY5PpYqjvS7J-y:-L6vwmRY5PpYqjvS7J-z:b-k2r3de) ([raw file](https://github.com/stratumn/go-indigocore/blob/390d13ae61c452d70ba949191b3cfe7f15e0555e/tmpop/tmpop.go#L376)):*
> ```Go
> 	linkHashes, err := t.getCommitLinkHashes(evidenceHeight)
> 	if err != nil {
> 		log.Warn("Could not get link hashes for this block. Evidence will not be generated.")
> ```

we should add more information about the block in the warning message

---

*[tmpop/tmpop.go, line 386 at r1](https://beta.reviewable.io/reviews/stratumn/go-indigocore/356#-L6vwqeMD6ZDTd_19mdn:-L6vwqeMD6ZDTd_19mdo:bpgwxg1) ([raw file](https://github.com/stratumn/go-indigocore/blob/390d13ae61c452d70ba949191b3cfe7f15e0555e/tmpop/tmpop.go#L386)):*
> ```Go
> 	validatorHash, err := t.getValidatorHash(evidenceHeight)
> 	if err != nil {
> 		log.Warn("Could not get validator hash for this block. Evidence will not be generated.")
> ```

here as well

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/stratumn/go-indigocore/356#-:-L6vyIYa2jFkbDU323bz:bnfp4nl)*

**Comment by such** [comment]
at last we have proper evidence! good job!

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/stratumn/go-indigocore/356#-:-L6vyNbx57znJgzHAyYn:ba28atl)*

**Comment by t-bast** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[cs/evidence_tendermint_test.go, line 189 at r1](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-L6vuIIG3kW5pP9uI1Oa:-L6vybIaWGEz0HHTW5KO:b-1dody0) ([raw file](https://github.com/stratumn/go-indigocore/blob/390d13ae61c452d70ba949191b3cfe7f15e0555e/cs/evidence_tendermint_test.go#L189)):*
<details><summary><i>Previously, such (Adrien Montfort) wrote…</i></summary><blockquote>

supernit: shouldn't this assert be in its own test case?
</blockquote></details>

I liked making sure that the proof was valid before I was passing it on to the tests that might modify it.
But now that it works, maybe it's not necessary, I'll have a second look at it.

---

*[tmpop/tmpop.go, line 376 at r1](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-L6vwmRY5PpYqjvS7J-y:-L6vys843Ivg2vKTtYmH:b-ovwaw7) ([raw file](https://github.com/stratumn/go-indigocore/blob/390d13ae61c452d70ba949191b3cfe7f15e0555e/tmpop/tmpop.go#L376)):*
<details><summary><i>Previously, such (Adrien Montfort) wrote…</i></summary><blockquote>

we should add more information about the block in the warning message
</blockquote></details>

Good idea.

---

*[tmpop/tmpop.go, line 386 at r1](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-L6vwqeMD6ZDTd_19mdn:-L6vytGKEWejWA4XxgAD:bfyf4eo) ([raw file](https://github.com/stratumn/go-indigocore/blob/390d13ae61c452d70ba949191b3cfe7f15e0555e/tmpop/tmpop.go#L386)):*
<details><summary><i>Previously, such (Adrien Montfort) wrote…</i></summary><blockquote>

here as well
</blockquote></details>

Good idea as well :).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/stratumn/go-indigocore/356)*

**Comment by t-bast** [comment]
Note that I'd like to work in the next few weeks on setting up a real metrics/monitoring stack for Indigo, this will be the occasion to provide good analytics on what happens in TMPoP.

---

Review status: 8 of 10 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-:-L6w1oAaFSGxzCpUC9k8:b3qkvgu)*

**Comment by such** [comment]
And a first brick for IndigoEntreprise. Nice!

---

Review status: 8 of 10 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-:-L6w2l2Mr4N_Udfs2wkx:bxdctua)*

**Comment by alexppxela** [comment]
Reviewed 9 of 10 files at r1, 1 of 2 files at r2.
Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[tmpop/tmpoptestcases/evidence.go, line 150 at r1](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-L6w7axYNlsB2M_Mtj50:-L6w7axYNlsB2M_Mtj51:brvsybb) ([raw file](https://github.com/stratumn/go-indigocore/blob/390d13ae61c452d70ba949191b3cfe7f15e0555e/tmpop/tmpoptestcases/evidence.go#L150)):*
> ```Go
> 	}
> 	tmClientMock.EXPECT().Block(int64(5)).Return(blocks[5], nil).AnyTimes()
> ```

nit: can factorize all block initialization

---

*[tmpop/tmpoptestcases/evidence.go, line 310 at r1](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-L6w86by41CHFtgjmbcI:-L6w86by41CHFtgjmbcJ:b2prq1g) ([raw file](https://github.com/stratumn/go-indigocore/blob/390d13ae61c452d70ba949191b3cfe7f15e0555e/tmpop/tmpoptestcases/evidence.go#L310)):*
> ```Go
> 		err := makeQuery(h, tmpop.GetSegment, linkHash4, got)
> 		assert.NoError(t, err)
> 		assert.Len(
> ```

nit: `assert.Empty()`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/stratumn/go-indigocore/356)*

**Comment by t-bast** [comment]
Yeah !

---

Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[tmpop/tmpoptestcases/evidence.go, line 150 at r1](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-L6w7axYNlsB2M_Mtj50:-L6wH5tMwzpQ2w1T7YhV:bfu7bdi) ([raw file](https://github.com/stratumn/go-indigocore/blob/390d13ae61c452d70ba949191b3cfe7f15e0555e/tmpop/tmpoptestcases/evidence.go#L150)):*
<details><summary><i>Previously, alexppxela (Alexandre Thibault) wrote…</i></summary><blockquote>

nit: can factorize all block initialization
</blockquote></details>

Not really, because a few blocks aren't built exactly like the others (some don't have votes).
This E2E test is big and a bit hard to follow I'll admit :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-:-L6wH3kd2E0KWXigQ_z3:b-x0zbxj)*

**Comment by t-bast** [comment]
Review status: 8 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[tmpop/tmpoptestcases/evidence.go, line 310 at r1](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-L6w86by41CHFtgjmbcI:-L6wHmyAHndthfa75ZfE:b-896fix) ([raw file](https://github.com/stratumn/go-indigocore/blob/390d13ae61c452d70ba949191b3cfe7f15e0555e/tmpop/tmpoptestcases/evidence.go#L310)):*
<details><summary><i>Previously, alexppxela (Alexandre Thibault) wrote…</i></summary><blockquote>

nit: `assert.Empty()`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/stratumn/go-indigocore/356)*

**Comment by alexppxela** [comment]
:lgtm:

---

Reviewed 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-:-L6zceW-rJXH17CDTLWS:bnfp4nl)*

**Comment by simonvadee** [comment]
looking great ! 

---

Reviewed 7 of 10 files at r1, 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[cs/evidence_tendermint_test.go, line 171 at r3](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-L6zbezoolbOSx8vLSMe:-L6zbezoolbOSx8vLSMf:b-8htpa9) ([raw file](https://github.com/stratumn/go-indigocore/blob/6b21010971d21cfbe660cee37ee3729f050b8faa/cs/evidence_tendermint_test.go#L171)):*
> ```Go
> // generates a valid block and its proof, and returns the link
> // and the evidence.
> func CreateTendermintProof(t *testing.T, linksCount int) (*types.Bytes32, *evidences.TendermintProof) {
> ```

nit: does this need to be exported ?

---

*[cs/evidences/evidences.go, line 192 at r2](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-L6wJcwtPWqsVtaafERy:-L6wJcwtPWqsVtaafERz:b4e7ubm) ([raw file](https://github.com/stratumn/go-indigocore/blob/2aab4cb1fc9b20c0919c8cdf873bb3e035687534/cs/evidences/evidences.go#L192)):*
> ```Go
> 
> 	// We validate that nodes signed the header.
> 	if !p.validateVotes(p.Header, p.HeaderVotes) {
> ```

as discussed IRL: are the (potential) byzantine votes included in `p.HeaderVotes` by tendermint ? if yes, then we should make sure than more than 1/3 of the votes are valid, not all of them.

---

*[tmpop/tmClient.go, line 86 at r3](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-L6zcgU3205olmLDsflN:-L6zcgU3205olmLDsflO:b-os4te0) ([raw file](https://github.com/stratumn/go-indigocore/blob/6b21010971d21cfbe660cee37ee3729f050b8faa/tmpop/tmClient.go#L86)):*
> ```Go
> 	for _, tx := range tmBlock.Block.Txs {
> 		tmTx, err := unmarshallTx(tx)
> 		if !err.IsOK() || tmTx.TxType != CreateLink {
> ```

why do you check if the transaction was a `CreateLink` (even though we only have this type) ?

---

*[tmpop/tmpoptestcases/evidence.go, line 337 at r3](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-L6zf9T4SkE3mQj3olCf:-L6zf9T4SkE3mQj3olCg:b2cv1g8) ([raw file](https://github.com/stratumn/go-indigocore/blob/6b21010971d21cfbe660cee37ee3729f050b8faa/tmpop/tmpoptestcases/evidence.go#L337)):*
> ```Go
> // vote creates a valid vote for a given header.
> // It simulates nodes signing a header and is crucial for the proof.
> func vote(header *tmtypes.Header) []*evidences.TendermintVote {
> ```

is this the same function as above (cs/evidence_tendermint_test.go) ? if yes, is there a way to factorize ? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-:-L6zffS0PFukxWYfJO5G:bif4nqs)*

**Comment by such** [comment]
The TendermintProof verification should be updated in JS now! ;)

---

Reviewed 1 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/stratumn/go-indigocore/356#-:-L6zgBZnBMzljEfIw7ru:bko03mb)*

**Comment by t-bast** [comment]
The developer named Bastien is not available anymore. Please try another developer.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[cs/evidence_tendermint_test.go, line 171 at r3](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-L6zbezoolbOSx8vLSMe:-L6zfpc0jTf147L02Jew:bjxnx24) ([raw file](https://github.com/stratumn/go-indigocore/blob/6b21010971d21cfbe660cee37ee3729f050b8faa/cs/evidence_tendermint_test.go#L171)):*
<details><summary><i>Previously, simonvadee (Simon Vadée) wrote…</i></summary><blockquote>

nit: does this need to be exported ?
</blockquote></details>

I think it's useful yes

---

*[cs/evidences/evidences.go, line 192 at r2](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-L6wJcwtPWqsVtaafERy:-L6zfrftwu9EXJyNg_ae:bqjrr76) ([raw file](https://github.com/stratumn/go-indigocore/blob/2aab4cb1fc9b20c0919c8cdf873bb3e035687534/cs/evidences/evidences.go#L192)):*
<details><summary><i>Previously, simonvadee (Simon Vadée) wrote…</i></summary><blockquote>

as discussed IRL: are the (potential) byzantine votes included in `p.HeaderVotes` by tendermint ? if yes, then we should make sure than more than 1/3 of the votes are valid, not all of them.
</blockquote></details>

Yes good point! I'll dive more into simulating byzantine nodes next.

---

*[tmpop/tmClient.go, line 86 at r3](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-L6zcgU3205olmLDsflN:-L6zfyZG_PHhiSyiqYYW:b-s2rx36) ([raw file](https://github.com/stratumn/go-indigocore/blob/6b21010971d21cfbe660cee37ee3729f050b8faa/tmpop/tmClient.go#L86)):*
<details><summary><i>Previously, simonvadee (Simon Vadée) wrote…</i></summary><blockquote>

why do you check if the transaction was a `CreateLink` (even though we only have this type) ?
</blockquote></details>

It feels more future-proof when/if we add other TxTypes :)
We most likely only want to create evidence for CreateLink operations (even though I admit it depends on what other operations we add in the future).
But the main usecase was that we might at some point have an AddEvidence operation to store external evidence, and we don't want to generate evidence on evidence...but that's still a bit blurry so we'll see later. 

---

*[tmpop/tmpoptestcases/evidence.go, line 337 at r3](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-L6zf9T4SkE3mQj3olCf:-L6zgI3GXb-gk7R1ipgM:b-cwo2rv) ([raw file](https://github.com/stratumn/go-indigocore/blob/6b21010971d21cfbe660cee37ee3729f050b8faa/tmpop/tmpoptestcases/evidence.go#L337)):*
<details><summary><i>Previously, simonvadee (Simon Vadée) wrote…</i></summary><blockquote>

is this the same function as above (cs/evidence_tendermint_test.go) ? if yes, is there a way to factorize ? 
</blockquote></details>

Yes it is, but I'm not a big fan of factorizing it yet, as Rob Pike says in Go sometimes a little duplication is better...if more functions need to be shared then we'll reevaluate :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/stratumn/go-indigocore/356#-:-L6zgVL7tiT52ywDuoZj:b-twa3cs)*

**Comment by such** [comment]
*[cs/evidences/evidences.go, line 192 at r2](https://beta.reviewable.io/reviews/stratumn/go-indigocore/356#-L6wJcwtPWqsVtaafERy:-L6zgUzS8hzktuFzhWUT:b-q9hcw9) ([raw file](https://github.com/stratumn/go-indigocore/blob/2aab4cb1fc9b20c0919c8cdf873bb3e035687534/cs/evidences/evidences.go#L192)):*
<details><summary><i>Previously, t-bast (Bastien Teinturier) wrote…</i></summary><blockquote>

Yes good point! I'll dive more into simulating byzantine nodes next.
</blockquote></details>

That's a bit different than byzantine votes. A wrong signature is a bad way to try to mess with the system since it's so easily catchable. I'm pretty sure bad signature are not included in the block header by the Tendermint engine.
A byzantine vote would be voting twice on blocks with the same height for instance.

But yeah we should check that we have at least 2/3+ signatures.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/stratumn/go-indigocore/356)*

**Comment by such** [comment]
^^

---







*Comments from [Reviewable](https://beta.reviewable.io/reviews/stratumn/go-indigocore/356#-:-L6zhSUYE4zba6EdnS9S:b32wed)*
